Revise DriveSubsystem

FireShredder24 opened the pull request:
Migrate the new DriveSubsystem back to master.
The new DriveSubsystem leverages the new DifferentialDriveKinematics class, the TalonFX velocity control mode, and the TalonFX built-in PID controller to provide the best driving experience Momentum has ever seen.

FireShredder24 committed:
Prepped for use of TalonFX PID

Deleted all aspects related to DifferentialDrive and WPILib PID

FireShredder24 committed:
Implement TalonFX VelocityControl

Use DifferentialDriveKinematics and TalonFX ControlMode.Velocity instead of DifferentialDrive and WPILib PID for the DriveSubsystem

FireShredder24 committed:
Implement Speed Limit Using Ternary in DriveSubsystem

jordan-powers commented on src/main/java/frc/robot/subsystems/DriveSubsystem.java:
```diff
@@ -11,103 +11,80 @@
   private final SensorCollection leftEnc = new SensorCollection(leftFront);
   private final SensorCollection rightEnc = new SensorCollection(rightFront);
 
+  private final DifferentialDriveKinematics kinematics = new DifferentialDriveKinematics(Units.inchesToMeters(26));
```
What's this 26?
Pull it out into a constant with a descriptive name, like you did with the ENC_TICKS_PER_FOOT

jordan-powers commented on src/main/java/frc/robot/subsystems/DriveSubsystem.java:
```diff
@@ -11,103 +11,80 @@
 
+  private final DifferentialDriveKinematics kinematics = new DifferentialDriveKinematics(Units.inchesToMeters(26));
+
   private final boolean pidEnabled = false;
```
This isn't used for anything anymore

jordan-powers commented on src/main/java/frc/robot/subsystems/DriveSubsystem.java:
```diff
@@ -11,103 +11,80 @@
-    drive.arcadeDrive(m_r, t_r, false);
-    // }
+  public void drive(double moveRequest, double turnRequest) {
+    var chassisSpeeds = new ChassisSpeeds(moveRequest, 0, turnRequest);
```
Hmm, what units are you expecting for moveRequest and turnRequest?
From the docs, ChassisSpeeds uses meters per second for its moveRequests and radians per second for its turnRequests.
Maybe add a javadoc comment to `drive()` explaining what the units of moveRequest and turnRequest need to be.

If you're expecting them to be between -1 and 1, you need to scale them by the maximum linear and rotational velocities of the robot before feeding them into the ChassisSpeeds

FireShredder24 replied:
I'm expecting a value between -1 and 1 from the Xbox controller joysticks.  I'll use map() to scale them correctly.

jordan-powers commented on src/main/java/frc/robot/subsystems/DriveSubsystem.java:
```diff
@@ -11,103 +11,80 @@
+    double leftMPerS = wheelSpeeds.leftMetersPerSecond;
+    double rightMPerS = wheelSpeeds.rightMetersPerSecond;
+    double leftETPerS = FeetToEncTicks(
+        Units.metersToFeet(leftMPerS) < 12 && Units.metersToFeet(leftMPerS) > -12 ? Units.metersToFeet(leftMPerS) : 12);
```
1. You have a magic number. Document 12 with a constant
2. You are clipping leftMPerS between -12 and 12. You should look at the [Utils.clip()](https://github.com/momentumfrc/2020InfiniteRecharge/blob/master/src/main/java/frc/robot/utils/Utils.java#L4-L6) method
3. I was under the impression that 12ft/s was the maximum linear velocity of the robot (moveRequest), not the maximum velocity of the wheel. I'm not sure though, so that's a question for @terbos90803

FireShredder24 replied:
While I was writing this, I considered using clip(), but I wanted to get practice using ternary notation.  There's no other reason why I wrote out that logic.

jordan-powers commented on src/main/java/frc/robot/subsystems/DriveSubsystem.java:
```diff
@@ -11,103 +11,80 @@
+    return ft * ENC_TICKS_PER_FOOT;
+  }
 
   public double getMeasurement() {
```
What use does this method serve?

FireShredder24 replied:
This method is a layer of indirection suggested by @terbos90803 that converts feet into TalonFX encoder ticks.  Its main purpose is to remove magic numbers from drive().

jordan-powers replied:
No, I was talking about getMeasurement()

terbos90803 replied:
Bump.  Question still stands.

FireShredder24 replied:
It was a carry-over from the previous WPILib PIDController-based approach, I just removed it.

jordan-powers commented on src/main/java/frc/robot/subsystems/DriveSubsystem.java:
```diff
@@ -11,103 +11,80 @@
-    drive = new DifferentialDrive(leftFront, rightFront);
-
-    drive.setDeadband(0);
   }
```
Hmm, you don't ever initialize any of the PID constants.
At the very least have kP, kI, kD, kIz, and kFF constants which you then set on the motors using `config_kP()`, `config_kI()`, `config_kD()`, `config_IntegralZone()`, and `config_kF()`.

See [here](https://phoenix-documentation.readthedocs.io/en/latest/ch16_ClosedLoop.html#velocity-closed-loop-control-mode) for tips on how to start finding kP, kI, and kD, and [here](https://phoenix-documentation.readthedocs.io/en/latest/ch16_ClosedLoop.html#velocity-closed-loop-control-mode) to see how to calculate a starting value for kF

jordan-powers commented on src/main/java/frc/robot/subsystems/DriveSubsystem.java:
```diff
@@ -11,103 +11,80 @@
+    double rightETPerS = FeetToEncTicks(
+        Units.metersToFeet(rightMPerS) < 12 && Units.metersToFeet(rightMPerS) > -12 ? Units.metersToFeet(leftMPerS)
+            : 12);
+    leftFront.set(ControlMode.Velocity, leftETPerS);
```
If you take in moveRequest and turnRequest as doubles between -1 and 1, then you can use an if statement to disable pid.
```
if(pidEnabled) {
  leftFront.set(ControlMode.Velocity, leftETPerS);
  rightFront.set(ControlMode.Velocity, rightETPerS);
} else {
  leftFront.set(ControlMode.PercentOutput, Utils.clip(moveRequest + turnRequest, -1, 1));
  rightFront.set(ControlMode.PercentOutput, Utils.clip(moveRequest - turnRequest, -1, 1));
}
```

jordan-powers commented on src/main/java/frc/robot/subsystems/DriveSubsystem.java:
```diff
@@ -11,103 +11,80 @@
-  public void useOutput(double output, double setpoint) {
-
-  };
+  private double FeetToEncTicks(double ft) {
```
Java conventions dictate that class names start with an upper case letter and class methods start with a lower case letter.

FireShredder24 committed:
WIP on Jordan's revisions

FireShredder24 committed:
TalonFX integrated PID configs

FireShredder24 committed:
Merge branch 'master' into driveSubsystem

jordan-powers commented:
Looks fine to me

terbos90803 commented on src/main/java/frc/robot/subsystems/DriveSubsystem.java:
```diff
@@ -11,103 +11,173 @@
+   * @param DRIVE_BASE_WIDTH_INCHES The number of inches between wheels on the
+   *                                drive base of the robot.
+   */
+  private final double DRIVE_BASE_WIDTH_INCHES = 26;
```
This constant describes the construction of the robot.  Things like this typically go in Constants.java, although you could argue that this is Main Drive specific.

Some of the other constants in this section should be evaluated the same way.  Anything that is clearly Drive related can stay here, but anything that could be referenced globally should move.  Perhaps this could be a refactoring activity.  i.e. Leave them here until they're needed elsewhere, then refactor at that time.

terbos90803 commented on src/main/java/frc/robot/subsystems/DriveSubsystem.java:
```diff
@@ -11,103 +11,173 @@
-    // Gets the acutal speed difference between left and right sides of the robot.
-    double turnRate = getTurnRate();
-
+    final double moveReqScaled = org.usfirst.frc.team4999.utils.Utils.map(moveRequest, -1, 1,
```
Consider importing Utils to shorten this line.

terbos90803 commented on src/main/java/frc/robot/subsystems/DriveSubsystem.java:
```diff
@@ -11,103 +11,173 @@
-    double turnRate = getTurnRate();
-
+    final double moveReqScaled = org.usfirst.frc.team4999.utils.Utils.map(moveRequest, -1, 1,
+        -Units.feetToMeters(SPEED_LIMIT_FEET_PER_S), Units.feetToMeters(SPEED_LIMIT_FEET_PER_S));
```
DRY principle.  Call the conversion once and use the result twice.

terbos90803 commented on src/main/java/frc/robot/subsystems/DriveSubsystem.java:
```diff
@@ -11,103 +11,173 @@
+   *                           second.
+   * 
+   */
+  private final double ENC_TICKS_PER_FOOT = 1304;
```
Consider converting to SI units to match kinematics class.

terbos90803 commented on src/main/java/frc/robot/subsystems/DriveSubsystem.java:
```diff
@@ -11,103 +11,173 @@
+      leftFront.set(ControlMode.Velocity, leftETPerS);
+      rightFront.set(ControlMode.Velocity, rightETPerS);
+    } else {
+      leftFront.set(ControlMode.PercentOutput, Utils.clip(moveRequest + turnRequest, -1, 1));
```
Since this mode ignores all the kinematics work, consider making this choice earlier and skipping the kinematics work entirely.

FireShredder24 committed:
Complete Terry's Reviews

FireShredder24 committed:
Complete Terry's requests part 2

terbos90803 commented on src/main/java/frc/robot/subsystems/DriveSubsystem.java:
```diff
@@ -105,43 +105,42 @@ public DriveSubsystem() {
+       * feed to ChassisSpeeds, we scale the moveRequest to the speed limit as
+       * converted to ft/s.
+       */
+      final double speedLimitMs = Units.feetToMeters(SPEED_LIMIT_FEET_PER_S);
```
If we're choosing SI units, be consistent and thorough about it.  Calling a method to convert a constant on every loop is wasteful.

FireShredder24 committed:
Remove wasteful conversion inside loop

States the value of the speed limit in meters per second instead of feet per second to remove wasteful conversion inside the closed-loop.

FireShredder24 committed:
Lower dangerously high speed limit

terbos90803 reviewed (approved):
Looks good.  Ship It.
